https://gcc.gnu.org/bugzilla/show_bug.cgi?id=122536#c4
--- a/gcc/simplify-rtx.cc
+++ b/gcc/simplify-rtx.cc
@@ -3686,8 +3686,8 @@ simplify_context::simplify_binary_operation_1 (rtx_code code,
 	  && XEXP (op0, 1) == CONST1_RTX (mode)
 	  && XEXP (op1, 1) == CONST1_RTX (mode)
 	  /* Verify bit positions (for cases with variable bit position). */
-	  && CONST_INT_P (XEXP (op0, 1))
-	  && CONST_INT_P (XEXP (op1, 1)))
+	  && CONST_INT_P (XEXP (XEXP (op0, 0), 1))
+	  && CONST_INT_P (XEXP (XEXP (op1, 0), 1)))
 	{
 	  unsigned HOST_WIDE_INT bitpos1 = INTVAL (XEXP (XEXP (op0, 0), 1));
 	  unsigned HOST_WIDE_INT bitpos2 = INTVAL (XEXP (XEXP (op1, 0), 1));
@@ -3718,7 +3718,7 @@ simplify_context::simplify_binary_operation_1 (rtx_code code,
 	  && XEXP (op0, 1) == CONST1_RTX (mode)
 	  && XEXP (op1, 1) == CONST0_RTX (mode)
 	  /* Verify bit position. */
-	  && CONST_INT_P (XEXP (op0, 1)))
+	  && CONST_INT_P (XEXP (XEXP (op0, 0), 1)))
 	{
 	  unsigned HOST_WIDE_INT bitpos1 = INTVAL (XEXP (XEXP (op0, 0), 1));
 	  unsigned HOST_WIDE_INT mask
